ENH: support markups with distorted zoom

kerim371 opened the pull request:
Fix markups representation when camera non-equally scaled along different axes.
Markups is designed to be always oriented away from camera.
Orientation is done with `GlyphOrientationArray`.
Scaling is designed to be similar to orientation.
With Glyph Mapper `SetScaleModeToScaleByVectorComponents`.
As orientation is updated with `UpdateControlPointGlyphOrientation`,
Scaling is updated with `UpdateControlPointGlyphScale` function. Anytime camera is changed `UpdateControlPointGlyphScale` is called.

Now it is a draft as I can't correctly project camera scaling from global coords to markups point coordinates.

**If you feel comfortable with coordinate transformation I would appreciate for any help (see `vtkSlicerMarkupsWidgetRepresentation3D::UpdateControlPointGlyphScale` function).**

To test markups with distorted camera scaling I usually do the following:
```python
layoutManager = slicer.app.layoutManager()
view = layoutManager.threeDWidget(0).threeDView()
threeDViewNode = view.mrmlViewNode()
cameraNode = slicer.modules.cameras.logic().GetViewActiveCameraNode(threeDViewNode)
camera=cameraNode.GetCamera()

# Scale Z-axis
transform = vtk.vtkTransform()
transform.Scale(1, 1, 2)
camera.SetModelTransformMatrix(transform.GetMatrix())
```

[See discussion](https://discourse.slicer.org/t/markups-distorted-when-zoom-isnt-constant-along-different-axes/24879)

kerim371 committed:
ENH: support markups with distorted zoom

Fix markups repr-n when camera non-equally scaled along different axes.
Markups is designed to be always oriented away from camera.
Orientation is done with `GlyphOrientationArray`.
Scaling is designed to be similar to orientation.
With Glyph Mapper `SetScaleModeToScaleByVectorComponents`.
As orientation is updated with `UpdateControlPointGlyphOrientation`,
Scaling is updated with `UpdateControlPointGlyphScale`.
Now it is a draft as I can't correctly project scaling from
global coords to markups point coordinates.

lassoan commented:
There are two markup sizing modes. If you choose absolute size then camera scaling should turn the spheres to ellipsoids. However, when sizing is relative then it could make sense to keep control points spherical. There may be similar issues with all markup types.

It is important not to impact performance of the most common use case (when camera is not scaled). If you add an extra filter or enable some more complicated option then check with profiling that speed is not impacted significantly.

kerim371 committed:
ENH: distortion transform sphere to ellipsoid

To compute scale that affect oriented markups we consider ellipsoid.
Semi-major/minor axes are camera scale.
Still this doesn't fully solve the problem.
When camera rolls markups may become distorted.

kerim371 commented:
Camera scale transforms sphere to ellipsoid in global coordinates.
To find scale affecting oriented markups we need to find a rotated point on that ellipsoid and compute radius vector.
Thus i,j,k vectors ({1,0,0}, {0,1,0}, {0,0,1}) after rotation changes their coordinates.
To compute `theta`, `phi`, we use spherical coordinates on rotated i,j,k (or we could rotate i,j,k vectors multiplied by scale and directly compute `theta`, `phi` from elliptical equations).
After that we can compute radius vector to the point on the ellipse. This will be the scale affecting rotated markups.

Anyway it seems VTK uses some logic that makes it impossible to compensate distortion with oriented object (see gif). Or mabe I'm missing something.

Links:
[Spherical coordinate system converssions](https://en.wikipedia.org/wiki/Spherical_coordinate_system#Coordinate_system_conversions)
[Ellipsoid parametrization](https://en.wikipedia.org/wiki/Ellipsoid#Parameterization)

![markups_with_inequal_zoom_ellipsoid](https://user-images.githubusercontent.com/43808863/187158817-b0d413a3-5d90-403a-a2eb-20e13cb5308c.gif)

lassoan commented:
Since you can apply a transform to each sphere, it must be possible to preserve spherical control point shape with any linear camera transform (there might be some slight residual distortion in perspective projection, but probably not).
You can test/tune your transform computation by displaying a simple cube model (first in the center and then in any position) and adjust the transform parameters to get a non-distorted cube.

kerim371 commented:
Yes, it should be doable using `vtkTransform` class.
I have tried this approach i.e. add `vtkTransformPolyDataFilter` and add `vtkSphereSource` to it.
But it seems it is not so easy to find the dependency of transformation matrix with orientation of markups. 
For example setting constant scales to `vtkTransform` lead to incorrect result as it seems it is applied to the local markups coordinate system as well as scale array applied to glyph mapper (i.e. it takes into account markups orientation):
![markups_with_inequal_zoom_constant_zoom_vtkTransform](https://user-images.githubusercontent.com/43808863/187264446-5f650c2f-49d5-4bd6-9188-3e04109e13aa.gif)

Only if I disable markups orientation updates then constant scales works well:
![markups_with_inequal_zoom_constant_zoom_vtkTransform_no_orientation](https://user-images.githubusercontent.com/43808863/187265344-dfb791dd-a688-4668-af5f-1e230f0b1a9a.gif)

Non-constant scales computed as described in ellipsoid section from previous post and applied to `vtkTransform` as scales also work not perfectly (most likely here the orientation is applied incorrectly - it is difficult to understand whether the rotation using markups orientation is correct or not):
![markups_with_inequal_zoom_vtkTransform](https://user-images.githubusercontent.com/43808863/187266781-6100326b-4541-46fa-b3bc-06c7049c7e09.gif)

lassoan commented:
Have you figured out the math for a simple model? Can you compute a transform that you can apply to a cube model at any position to get it displayed at the correct position with the correct shape?

kerim371 commented:
> Have you figured out the math for a simple model? Can you compute a transform that you can apply to a cube model at any position to get it displayed at the correct position with the correct shape?

Yes I tried that only with non-oriented simple cube (outside Slicer). It works perfectly (see gif and code below: here the zoom along Z-axis is `2`, and I apply scale to the cube in Z-direction equal to `0.5` and the cube stays the cube with origin `[2.5, 2.5, 2.5]`).

The problems starts when the object is oriented i.e. when `this->GlyphMapper->OrientOn();` within Slicer markups module.
The math becomes complicated and I'm not sure that I understand it correctly. Mostly I'm not sure whether I rotate points in correct direction (I tried both using `orientationQuaternion.Invert();` if this really changes the direction of rotation).

As far as I know linear transformation includes only: translation, rotation and scaling. Superposition of that is all I have. If I scale only the camera then I need to scale only the object.

![simple_cube](https://user-images.githubusercontent.com/43808863/187273025-1862153a-a18b-46c3-8e10-a5c404926023.gif)
```c++
#include <vtkActor.h>
#include <vtkAppendFilter.h>
#include <vtkDataSetMapper.h>
#include <vtkGlyph3DMapper.h>
#include <vtkNamedColors.h>
#include <vtkNew.h>
#include <vtkPointSource.h>
#include <vtkPoints.h>
#include <vtkPolyData.h>
#include <vtkProperty.h>
#include <vtkRenderWindow.h>
#include <vtkRenderWindowInteractor.h>
#include <vtkRenderer.h>
#include <vtkSmartPointer.h>
#include <vtkSphereSource.h>
#include <vtkCubeSource.h>
#include <vtkUnstructuredGrid.h>
#include <vtkTransform.h>
#include <vtkTransformPolyDataFilter.h>
#include <vtkCubeAxesActor.h>
#include <vtkProperty.h>
#include <vtkTextProperty.h>
#include <vtkCamera.h>
#include <vtkDoubleArray.h>
#include <vtkPointData.h>


int main(int, char*[])
{
  vtkNew<vtkCubeSource> cubeSource;
  cubeSource->SetXLength(1);
  cubeSource->SetYLength(1);
  cubeSource->SetZLength(1);

  vtkNew<vtkTransform> transform;
  transform->Scale(1,1,0.5);

  vtkNew<vtkTransformPolyDataFilter> transformFilter;
  transformFilter->SetInputConnection(cubeSource->GetOutputPort());
  transformFilter->SetTransform(transform);

  // Create 2 points in a vtkUnstructuredGrid
  vtkNew<vtkPoints> points;
  points->InsertNextPoint(2.5, 2.5, 2.5);

  vtkNew<vtkPolyData> polyData;
  polyData->SetPoints(points);

  // Not used when vtkTransform is in use
  vtkNew<vtkDoubleArray> scaleArr;
  scaleArr->SetName("scale");
  scaleArr->SetNumberOfComponents(3);
  scaleArr->SetNumberOfTuples(points->GetNumberOfPoints());
  scaleArr->SetTuple3(0, 1, 1, 0.5);
  polyData->GetPointData()->AddArray(scaleArr);

  vtkNew<vtkGlyph3DMapper> mapper;
  mapper->SetInputData(polyData);
  mapper->SetSourceConnection(transformFilter->GetOutputPort());
  mapper->ScalarVisibilityOff();
  // mapper->SetScaleModeToScaleByVectorComponents();
  // mapper->SetScaleArray("scale");
  // mapper->SetScaleFactor(1.0);

  // Create a mapper and actor
  vtkNew<vtkNamedColors> colors;

  // Glyph mapper
  vtkNew<vtkActor> actor;
  actor->SetMapper(mapper);
  // actor->SetPosition(0,0,3);

  // Create a renderer, render window, and interactor
  vtkNew<vtkRenderer> renderer;

  // Scale the camera
  vtkNew<vtkTransform> cameraTransform;
  cameraTransform->Scale(1,1,2);
  vtkCamera* camera = renderer->GetActiveCamera();
  camera->SetModelTransformMatrix(cameraTransform->GetMatrix());

  // Cube axes actor
  vtkNew<vtkCubeAxesActor> cubeAxesActor;
  cubeAxesActor->SetUseTextActor3D(1);
  cubeAxesActor->SetBounds(0,5,0,5,0,5);
  cubeAxesActor->SetCamera(renderer->GetActiveCamera());
  cubeAxesActor->GetTitleTextProperty(0)->SetColor(colors->GetColor3d("Salmon").GetData());
  cubeAxesActor->GetTitleTextProperty(0)->SetFontSize(48);
  cubeAxesActor->GetLabelTextProperty(0)->SetColor(colors->GetColor3d("Salmon").GetData());

  cubeAxesActor->GetTitleTextProperty(1)->SetColor(colors->GetColor3d("PaleGreen").GetData());
  cubeAxesActor->GetLabelTextProperty(1)->SetColor(colors->GetColor3d("PaleGreen").GetData());

  cubeAxesActor->GetTitleTextProperty(2)->SetColor(colors->GetColor3d("LightSkyBlue").GetData());
  cubeAxesActor->GetLabelTextProperty(2)->SetColor(colors->GetColor3d("LightSkyBlue").GetData());

  cubeAxesActor->DrawXGridlinesOn();
  cubeAxesActor->DrawYGridlinesOn();
  cubeAxesActor->DrawZGridlinesOn();
#if VTK_MAJOR_VERSION == 6
  cubeAxesActor->SetGridLineLocation(VTK_GRID_LINES_FURTHEST);
#endif
#if VTK_MAJOR_VERSION > 6
  cubeAxesActor->SetGridLineLocation(cubeAxesActor->VTK_GRID_LINES_FURTHEST);
#endif

  cubeAxesActor->XAxisMinorTickVisibilityOff();
  cubeAxesActor->YAxisMinorTickVisibilityOff();
  cubeAxesActor->ZAxisMinorTickVisibilityOff();

  cubeAxesActor->SetFlyModeToStaticEdges();

  // Add the actor to the scene
  renderer->AddActor(actor);
  renderer->AddActor(cubeAxesActor);
  renderer->GetActiveCamera()->Azimuth(30);
  renderer->GetActiveCamera()->Elevation(30);

  renderer->ResetCamera();
  renderer->SetBackground(colors->GetColor3d("RoyalBlue").GetData());

  vtkNew<vtkRenderWindow> renderWindow;
  renderWindow->AddRenderer(renderer);
  renderWindow->SetSize(640, 480);

  vtkNew<vtkRenderWindowInteractor> renderWindowInteractor;
  renderWindowInteractor->SetRenderWindow(renderWindow);

  // Render and interact
  renderWindow->SetWindowName("Scaled Glyph3D");
  renderWindow->Render();
  renderWindowInteractor->Start();

  return EXIT_SUCCESS;
}
```
